src/System.Private.Reflection.Core/src/System/Reflection/Runtime/TypeInfos/RuntimeTypeInfo.cs

```diff
@@ -192,9 +193,69 @@ public sealed override MemberInfo[] GetDefaultMembers()
             return defaultMemberName != null ? GetMember(defaultMemberName) : Array.Empty<MemberInfo>();
         }
 
-        public sealed override InterfaceMapping GetInterfaceMap(Type interfaceType)
+        public sealed override InterfaceMapping GetInterfaceMap(Type ifaceType)
+        {
+            if (IsGenericParameter)
+                throw new InvalidOperationException(SR.Arg_GenericParameter);
+
+            if (ifaceType is null)
+                throw new ArgumentNullException(nameof(ifaceType));
+
+            RuntimeTypeInfo ifaceRtType = ifaceType as RuntimeTypeInfo;
+
+            if (ifaceRtType == null)
+                throw new ArgumentException(SR.Argument_MustBeRuntimeType, nameof(ifaceType));
+
+            RuntimeTypeHandle ifaceRtTypeHandle = ifaceRtType.TypeHandle;
+
+            Debug.Assert(ifaceType.IsInterface);
+            Debug.Assert(!IsInterface);
+            VerifyInterfaceIsImplemented(ifaceRtTypeHandle);
+
+            // SZArrays implement the methods on IList`1, IEnumerable`1, and ICollection`1 with
+            // SZArrayHelper and some runtime magic. We don't have accurate interface maps for them.
+            if (IsSZArray && ifaceType.IsGenericType)
+                throw new ArgumentException(SR.Argument_ArrayGetInterfaceMap);
+
+            int ifaceInstanceMethodCount = RuntimeAugments.GetNumVirtuals(ifaceRtTypeHandle);
+
+            InterfaceMapping im;
+            im.InterfaceType = ifaceType;
+            im.TargetType = this;
+            im.InterfaceMethods = new MethodInfo[ifaceInstanceMethodCount];
+            im.TargetMethods = new MethodInfo[ifaceInstanceMethodCount];
+
+            var ifaceMethods = ifaceRtType.GetMethods();
```

Comment:
I couldn't find a way to map an interface slot number to the interface method (the VTable slots are there but all set to 0) and the methods returned by `GetMethods` appear to be in slot number order...

Reply:
The reflection policies currently in use in CoreRT should make it so that this is mostly true, but in general since reflection metadata is just an addon and not essential, it's possible to end up in a situation where a method has a slot and implementations, but we didn't generate metadata for it.

E.g. an interface like this:

```csharp
public interface IFoo
{
    internal void Frob();
    void Boink();
}
```

in an assembly that doesn't generate metadata for its privates (i.e. defines a `System.Runtime.CompilerServices.__BlockReflectionAttribute`) would only have a single method returned from `GetMethods()`, but still return 2 from `GetNumVirtuals`.

Similarly, an interface that has no methods actually used in code, but got compiled with the [IlcGenerateCompleteTypeMetadata](https://github.com/dotnet/corert/blob/66fbcd492fbc08db4f472e7e8fa368cb523b38d4/Documentation/using-corert/optimizing-corert.md), would return all the methods in `GetMethods`, but see a 0 in `ifaceInstanceMethodCount`.

Joys of optional metadata.

Maybe this is another of those "perfect being the enemy of the good enough". The actual fix for this is a bit more involved.

Reply:
So i was testing your problematic scenarios (i was testing with code that should produce the `see a 0 in ifaceInstanceMethodCount` scenario but didn't) and when i tried a generic interface it really broke down (`GetNumVirtuals` suddenly returned fewer methods and most of them were zero).

A night of sleep later i had another idea how to solve this so i copied the mechanism for how a `MethodInfo` of an interface can invoke a method on a concrete type, which deep down knows the mapping from an interface slot to the instance slot.

I didn't quite realize at first this was possible but it makes sense, although it does make me wonder what the special purpose of `GetInterfaceMap` is, e.g. when it is irreplaceable.

The new implementation is more reflection heavy but plays nice with the scenarios you've listed or throws this little bit confusing message:
```
Unhandled Exception: EETypeRva:0x00630940(System.Reflection.MissingRuntimeArtifactException): This object cannot be invoked because it was metadata-enabled for browsing only: 'ConsoleApp83.IInterface<System.Int32>.get_Value()' For more information, please visit  http://go.microsoft.com/fwlink/?LinkID=616867
```